src/main/java/com/commercetools/sync/products/service/impl/ProductReferenceTransformServiceImpl.java

```diff
@@ -0,0 +1,496 @@
+import com.commercetools.sync.commons.models.ResourceIdsGraphQlRequest;
+import com.commercetools.sync.commons.utils.ChunkUtils;
+import com.commercetools.sync.customobjects.helpers.CustomObjectCompositeIdentifier;
+import com.commercetools.sync.products.service.ProductReferenceTransformService;
+import com.commercetools.sync.services.impl.BaseTransformServiceImpl;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+import io.sphere.sdk.categories.Category;
+import io.sphere.sdk.client.SphereClient;
+import io.sphere.sdk.customobjects.CustomObject;
+import io.sphere.sdk.customobjects.queries.CustomObjectQuery;
+import io.sphere.sdk.models.Asset;
+import io.sphere.sdk.models.Reference;
+import io.sphere.sdk.products.AttributeContainer;
+import io.sphere.sdk.products.Price;
+import io.sphere.sdk.products.Product;
+import io.sphere.sdk.products.ProductDraft;
+import io.sphere.sdk.products.ProductLike;
+import io.sphere.sdk.products.ProductVariant;
+import io.sphere.sdk.products.attributes.Attribute;
+import io.sphere.sdk.producttypes.ProductType;
+import io.sphere.sdk.types.CustomFields;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.CompletionStage;
+import java.util.stream.Collectors;
+import javax.annotation.Nonnull;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class ProductReferenceTransformServiceImpl extends BaseTransformServiceImpl
+    implements ProductReferenceTransformService {
+
+  private static final Logger LOGGER =
+      LoggerFactory.getLogger(ProductReferenceTransformServiceImpl.class);
+  private static final String WITH_IRRESOLVABLE_REFS_ERROR_MSG =
+      "The product with id '%s' on the source project ('%s') will "
+          + "not be synced because it has the following reference attribute(s): %n"
+          + "%s.%nThese references are either pointing to a non-existent resource or to an existing one but with a blank key. "
+          + "Please make sure these referenced resources are existing and have non-blank (i.e. non-null and non-empty) keys.";
+
+  public ProductReferenceTransformServiceImpl(
+      @Nonnull final SphereClient ctpClient,
+      @Nonnull final Map<String, String> referenceIdToKeyCache) {
+    super(ctpClient, referenceIdToKeyCache);
+  }
+
+  @Nonnull
+  @Override
+  public CompletableFuture<List<ProductDraft>> transformProductReferences(
+      @Nonnull final List<Product> products) {
+
+    return replaceAttributeReferenceIdsWithKeys(products)
+        .handle(
+            (productsResolved, throwable) -> {
+              if (throwable != null && LOGGER.isWarnEnabled()) {
+                LOGGER.warn(
```

Comment:
Looks like this error handling part is more related to the project-sync code, not related to java-sync . In this case, we might throw the error from the service as we could not resolve this right? The library user should decide what to do.

Reply:
https://github.com/commercetools/commercetools-sync-java/pull/697/commits/c15173a3fe3479816c7abb1d34ecc09a498b8f0e